### ISSUE TITLE
Cam0 vs Cam1 as an input device - unexpected results

### ISSUE THREAD

**Issue by nsimon1** [6by9/raspiraw#8]
@6by9 Thank you for all your help so far, after [pointing me towards your yavta fork](https://github.com/waveform80/picamera/issues/481) I decided to keep working on making the pipeline in C instead of using picamera.  

I have found some strange results when using CAM0 vs CAM1 as the rawcam input component.

I am using the following setup for testing right now:  
Camera -> FPGA --> CSI-2 Interface [Continuous YUYV stream] --> CM3 CAM0/CAM1 input --> raspiraw (I2C disabled) --> ISP [YUYV to I420 conversion] --> Render

**Using CAM0 as the input (Perfect render image but green encoded output)**  
I started off using the CAM0 interface and I got the ISP conversion and render display working right away by using the functionality already in raspiraw.c, just changing the encoding, selecting cam 0 and specifying the correct resolution. The picture came out perfect. My end goal is rawcam->isp->encode->.h264 out and this is where I ran into problems. Whenever I  tunneled the isp output to either a splitter or encoder component instead of the render component the output would be a green screen.
[Link to video](https://streamable.com/qjy0z) of rawcam->isp->render working on CAM0  
[vcdbg log msg](https://github.com/6by9/raspiraw/files/1917592/msg_cam0.txt) for rawcam->isp->render

**Using CAM1 as the input (Sporadic render image but can now be encoded)**
After not getting the encoding or isp->splitter->render working I decided to try out CAM1 as the input, trying to stick to the default setup. This when unexpected things started happening. The previous rawcam->isp->render pipeline now resulted in a video output that was very sporadic or sometimes just froze right when launched with part of the image at the top of the frame and green below it. The strange thing is that once I took this input and tried to tunnel it to the splitter->render or encoder it worked! The result was the same sporadic video but now encoded.  

[Link to video](https://streamable.com/n2cc8) when connected via CAM1 and running `./raspiraw -t 5000 -w 1280 -h 720` in the case where it does stream.
[Image](https://ibb.co/hAvoK7) where it gets frozen right when the program starts.

[vcdbg log msg (streaming case)](https://github.com/6by9/raspiraw/files/1917594/msg_loop.txt) for rawcam->isp->render  
[vcdbg log msg (frozen case)](https://github.com/6by9/raspiraw/files/1917593/msg_frozen.txt) for rawcam->isp->render  

**General Notes**  

- I am using the latest firmware (rpi-update)

- `Linux raspberrypi 4.14.34-v7+ #1110 SMP Mon Apr 16 15:18:51 BST 2018 armv7l GNU/Linux`  

- `Raspbian GNU/Linux 9 (stretch)`

- `Board revsion: a220a0`

- There is no I2C communication between the RPi and the FPGA/Camera. The camera is in a continuous streaming mode upon power up. 

I am having a hard time understanding why there is a difference between the CAM0 and CAM1 attempts. I would appreciate any insights on what could be going on behind the scenes or additional tests that I can do to help narrow it down.  

In the case of the sporadic video, I have never seen this but I am hoping that someone else has, and could help shed some light on the issue.

**Comment by 6by9** [comment]
I've no immediate answers as to why CAM0 should give good images but not encode vs CAM1 which encodes but with rubbish images.
I did have an initial thought that rawcam spits out a buffer potentially containing the embedded data as well as the image data, and that could confuse some components. ISP specifically has handling for it in though, so that knocks that thought out of the water.
GPU mem would be my other thought, but it wouldn't be different between the two peripherals.

Truth be told I've never used CAM0 with either rawcam or the V4L2 driver. I do have a TC358743 hooked up to a CM3L at the moment, so I'll flick across and confirm what happens there.

**Comment by 6by9** [comment]
Can you post your code so that I can review it and confirm there isn't anything daft in the setup? Email it if you don't want to put it in the public domain (see email address on most of the commits in this repo).

**Comment by nsimon1** [comment]
**"Using CAM0 as the input (Perfect render image but green encoded output)"** is now solved!  
I got a B101 in today and was able to do some testing. I found out that I was missing the `rx_cfg.image_id` value required for the YUYV input data. When I changed the value in the tc358743 code from 0x24 to another ID I found that the buffers were only filling up with a few bytes as they would when I got the green output screen, except the output was black instead of green this time.  

I now have csi0 working through the whole pipeline as in the tc358743 code:  
```
rawcam --> isp --> splitter --> render
                            --> encoder --> output
```

- I am not sure why the `isp --> render` previously worked without the image_id at all though.  

- I am not sending any embedded data along with the image data and I have found that changing `rx_cfg.embedded_data_lines` between the `128` and `0` makes no difference and the video works either way. I don't know if not sending any embedded data might be related to the csi1 input issue. 

**"Using CAM1 as the input (Sporadic render image but can now be encoded)"** is still an issue.  

- I started off with a new clone of your tc358743 repo and made changes so that it works with CAM0 and produces the working results as above. [Here is that version](https://github.com/nsimon1/raspi_tc358743/blob/master/raspi_tc358743.c). I will strip out all the unnecessary parts tomorrow and I can email you a cleaned up version then, which might make things easier. 

- When I swap the FPGA output to csi1 instead of csi0 and just change the `cam_num` parameter then I end up with the sporadic video output. 

- I have also noted the following errors in the log while the sporadic video is streaming:  
```
039015.501: output type is FRAME
039015.645: unicam_int_callback: instance=1 status = 0x7010
039015.665: unicam_int_callback: bytes_written = 168960, lines_done = 66
039015.688: unicam_int_callback error: parity bit error
039015.705: unicam_int_callback error: single bit error error
039015.721: unicam_int_callback error: higher order error
039015.739: unicam_int_callback: Frame End
039015.755: cm_process_thread: events 0010
039015.766: cm_unicam_event - enter
039015.778: cm_unicam_event: FRAME_END complete
039015.797: cm_unicam_event - exit
``` 
[Whole vcdgb log](https://github.com/6by9/raspiraw/files/1921881/msg_csi1_2018-04-17.txt)


- I am not sure if there may be any other CSI related settings that are applied by default to csi1 vs csi0 that I might be missing?
